File: langs/ruby/lib/dotlyte/interpolation.rb
# frozen_string_literal: true

module Dotlyte
  # Variable interpolation engine for DOTLYTE v2.
  #
  # Supports ${VAR}, ${VAR:-default}, ${VAR:?error}, and $$ escape.
  module Interpolation
    # Interpolate ${VAR} references in a flat string hash.
    #
    # @param data [Hash<String, String>] key-value pairs
    # @param context [Hash<String, String>] additional context
    # @return [Hash<String, String>] resolved values
    def self.interpolate(data, context = {})
      resolved = {}
      resolving = Set.new

      data.each_key { |key| resolve(key, data, context, resolved, resolving) }
      resolved
    end

    # Interpolate a deep hash (nested values).
    #
    # @param data [Hash] deep hash
    # @param context [Hash] additional context
    # @return [Hash] resolved deep hash
    def self.interpolate_deep(data, context = {})
      flat = flatten_to_strings(data)
      ctx_flat = flatten_to_strings(context)
      resolved = interpolate(flat, ctx_flat)

      result = deep_copy(data)
      resolved.each do |key, value|
        set_nested(result, key, Coercion.coerce(value))
      end
      result
    end

    class << self
      private

      def resolve(key, data, context, resolved, resolving)
        return resolved[key] if resolved.key?(key)

        if resolving.include?(key)
          raise InterpolationError.new(
            "Circular reference detected for variable: #{key}", key: key
          )
        end

        raw = data[key]
        if raw.nil?
          return context[key] if context.key?(key)

          env = ENV[key.upcase]
          return env || ""
        end

        resolving.add(key)
        val = resolve_string(raw.to_s, data, context, resolved, resolving)
        resolving.delete(key)
        resolved[key] = val
        val
      end

      def resolve_string(s, data, context, resolved, resolving)
        s = s.gsub("$$", "\x00DOLLAR\x00")
        result = +""
        i = 0

        while i < s.length
          if i + 1 < s.length && s[i] == "$" && s[i + 1] == "{"
            i += 2
            depth = 1
            inner = +""
            while i < s.length && depth.positive?
              ch = s[i]
              if ch == "{"
                depth += 1
              elsif ch == "}"
                depth -= 1
                if depth.zero?
                  i += 1
                  break
                end
              end
              inner << ch
              i += 1
            end
            result << resolve_reference(inner, data, context, resolved, resolving)
          else
            result << s[i]
            i += 1
          end
        end

        result.gsub("\x00DOLLAR\x00", "$")
      end

      def resolve_reference(inner, data, context, resolved, resolving)
        err_idx = inner.index(":?")
        def_idx = inner.index(":-")

        if err_idx
          var_name = inner[0...err_idx].strip
          error_msg = inner[(err_idx + 2)..]
        elsif def_idx
          var_name = inner[0...def_idx].strip
          fallback = inner[(def_idx + 2)..]
        else
          var_name = inner.strip
        end

        lower = var_name.downcase

        # Same-file
        if data.key?(lower)
          val = resolve(lower, data, context, resolved, resolving)
          return val unless val.empty?
        end

        # Context
        if context.key?(lower)
          val = context[lower]
          return val if val && !val.empty?
        end

        # Env
        env = ENV[var_name] || ENV[var_name.upcase]
        return env if env && !env.empty?

        # Not found
        if error_msg
          raise Error.new("Required variable '#{var_name}': #{error_msg}", key: var_name)
        end

        fallback || ""
      end

      def flatten_to_strings(hash, prefix = "", out = {})
        hash.each do |key, value|
          full_key = prefix.empty? ? key.to_s : "#{prefix}.#{key}"
          if value.is_a?(Hash)
            flatten_to_strings(value, full_key, out)
          elsif !value.nil?
            out[full_key] = value.to_s
          end
        end
        out
      end

      def deep_copy(hash)
        hash.each_with_object({}) do |(k, v), result|
          result[k] = v.is_a?(Hash) ? deep_copy(v) : v
        end
      end

      def set_nested(hash, key, value)
        parts = key.split(".")
        current = hash
        parts[0..-2].each do |part|
          current[part] ||= {}
          current[part] = {} unless current[part].is_a?(Hash)
          current = current[part]
        end
        current[parts.last] = value
      end
    end
  end
end

File: langs/dotnet/src/Dotlyte/Interpolation.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Dotlyte;

/// <summary>
/// Variable interpolation engine for DOTLYTE v2.
/// Supports ${VAR}, ${VAR:-default}, ${VAR:?error}, and $$ escape.
/// </summary>
public static class Interpolation
{
    /// <summary>
    /// Interpolate ${VAR} references in a flat string dictionary.
    /// </summary>
    public static Dictionary<string, string> Interpolate(
        Dictionary<string, string> data,
        Dictionary<string, string>? context = null)
    {
        context ??= new Dictionary<string, string>();
        var resolved = new Dictionary<string, string>();
        var resolving = new HashSet<string>();

        foreach (var key in data.Keys)
        {
            Resolve(key, data, context, resolved, resolving);
        }

        return resolved;
    }

    /// <summary>
    /// Interpolate a deep (nested) dictionary.
    /// Only string values containing ${...} are interpolated.
    /// </summary>
    public static Dictionary<string, object?> InterpolateDeep(
        Dictionary<string, object?> data,
        Dictionary<string, object?>? context = null)
    {
        var flat = FlattenToStrings(data);
        var ctxFlat = context is not null ? FlattenToStrings(context) : new Dictionary<string, string>();

        var needsInterpolation = flat
            .Where(kvp => kvp.Value.Contains("${"))
            .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);

        if (needsInterpolation.Count == 0)
            return DeepCopy(data);

        var resolved = Interpolate(flat, ctxFlat);
        var result = DeepCopy(data);

        foreach (var key in needsInterpolation.Keys)
        {
            if (resolved.TryGetValue(key, out var val))
            {
                SetNested(result, key, Coercion.Coerce(val));
            }
        }

        return result;
    }

    private static string Resolve(
        string key,
        Dictionary<string, string> data,
        Dictionary<string, string> context,
        Dictionary<string, string> resolved,
        HashSet<string> resolving)
    {
        if (resolved.TryGetValue(key, out var cached))
            return cached;

        if (resolving.Contains(key))
            throw new InterpolationException(
                $"Circular reference detected for variable: {key}", key);

        if (!data.TryGetValue(key, out var rawValue))
        {
            if (context.TryGetValue(key, out var ctxVal))
                return ctxVal;
            var env = Environment.GetEnvironmentVariable(key.ToUpperInvariant());
            return env ?? string.Empty;
        }

        resolving.Add(key);
        var val2 = ResolveString(rawValue, data, context, resolved, resolving);
        resolving.Remove(key);
        resolved[key] = val2;
        return val2;
    }

    private static string ResolveString(
        string s,
        Dictionary<string, string> data,
        Dictionary<string, string> context,
        Dictionary<string, string> resolved,
        HashSet<string> resolving)
    {
        const string placeholder = "\x00DOLLAR\x00";
        s = s.Replace("$$", placeholder);

        var sb = new StringBuilder();
        var i = 0;

        while (i < s.Length)
        {
            if (i + 1 < s.Length && s[i] == '$' && s[i + 1] == '{')
            {
                i += 2;
                var depth = 1;
                var inner = new StringBuilder();
                while (i < s.Length && depth > 0)
                {
                    if (s[i] == '{') depth++;
                    else if (s[i] == '}')
                    {
                        depth--;
                        if (depth == 0) { i++; break; }
                    }
                    inner.Append(s[i]);
                    i++;
                }
                sb.Append(ResolveReference(inner.ToString(), data, context, resolved, resolving));
            }
            else
            {
                sb.Append(s[i]);
                i++;
            }
        }

        return sb.ToString().Replace(placeholder, "$");
    }

    private static string ResolveReference(
        string inner,
        Dictionary<string, string> data,
        Dictionary<string, string> context,
        Dictionary<string, string> resolved,
        HashSet<string> resolving)
    {
        string varName;
        string? errorMsg = null;
        string? fallback = null;

        var errIdx = inner.IndexOf(":?", StringComparison.Ordinal);
        var defIdx = inner.IndexOf(":-", StringComparison.Ordinal);

        if (errIdx >= 0)
        {
            varName = inner[..errIdx].Trim();
            errorMsg = inner[(errIdx + 2)..];
        }
        else if (defIdx >= 0)
        {
            varName = inner[..defIdx].Trim();
            fallback = inner[(defIdx + 2)..];
        }
        else
        {
            varName = inner.Trim();
        }

        var lower = varName.ToLowerInvariant();

        // Same-file
        if (data.ContainsKey(lower))
        {
            var val = Resolve(lower, data, context, resolved, resolving);
            if (!string.IsNullOrEmpty(val))
                return val;
        }

        // Context
        if (context.TryGetValue(lower, out var ctxVal) && !string.IsNullOrEmpty(ctxVal))
            return ctxVal;

        // Environment
        var env = Environment.GetEnvironmentVariable(varName)
            ?? Environment.GetEnvironmentVariable(varName.ToUpperInvariant());
        if (!string.IsNullOrEmpty(env))
            return env;

        // Not found
        if (errorMsg is not null)
            throw new InterpolationException(
                $"Required variable '{varName}': {errorMsg}", varName);

        return fallback ?? string.Empty;
    }

    private static Dictionary<string, string> FlattenToStrings(
        Dictionary<string, object?> data, string prefix = "")
    {
        var result = new Dictionary<string, string>();
        foreach (var (key, value) in data)
        {
            var fullKey = string.IsNullOrEmpty(prefix) ? key : $"{prefix}.{key}";
            if (value is Dictionary<string, object?> nested)
            {
                foreach (var (k, v) in FlattenToStrings(nested, fullKey))
                    result[k] = v;
            }
            else if (value is not null)
            {
                result[fullKey] = value.ToString()!;
            }
        }
        return result;
    }

    private static Dictionary<string, object?> DeepCopy(Dictionary<string, object?> data)
    {
        var result = new Dictionary<string, object?>();
        foreach (var (k, v) in data)
        {
            result[k] = v is Dictionary<string, object?> nested
                ? DeepCopy(nested)
                : v;
        }
        return result;
    }

    private static void SetNested(Dictionary<string, object?> data, string key, object? value)
    {
        var parts = key.Split('.');
        var current = data;
        for (var i = 0; i < parts.Length - 1; i++)
        {
            if (!current.TryGetValue(parts[i], out var existing) ||
                existing is not Dictionary<string, object?> dict)
            {
                dict = new Dictionary<string, object?>();
                current[parts[i]] = dict;
            }
            current = dict;
        }
        current[parts[^1]] = value;
    }
}
